### PR TITLE
GT-183 Pregel API support

### PR DESCRIPTION
**Pull request by jwierzbo**

### CODE REVIEW

File: database_pregel.go [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+	// Name of the algorithm
+	Algorithm PregelAlgorithm `json:"algorithm"`
+	// Name of a graph. Either this or the parameters VertexCollections and EdgeCollections are required.
+	// Please note that there are special sharding requirements for graphs in order to be used with Pregel.
+	GraphName string `json:"graphName,optional"`
+	// List of vertex collection names. Please note that there are special sharding requirements for collections in order to be used with Pregel.
+	VertexCollections []string `json:"vertexCollections,optional"`
+	// List of edge collection names. Please note that there are special sharding requirements for collections in order to be used with Pregel.
+	EdgeCollections []string `json:"edgeCollections,optional"`
+	// General as well as algorithm-specific options.
```
Comment (nikita-vanyasin) [inline database_pregel.go:65]:
```suggestion
	// General and algorithm-specific options.
```

File: database_pregel.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+	Params map[string]interface{} `json:"params,optional"`
+}
+
+type PregelJobState string
+
+const (
+	// PregelJobNone - The Pregel run did not yet start.
+	PregelJobNone PregelJobState = "none"
```
Comment (nikita-vanyasin) [inline database_pregel.go:73]:
```suggestion
	PregelJobStateNone PregelJobState = "none"
```

File: database_pregel.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+	// PregelLoading - The graph is loaded from the database into memory before the execution of the algorithm.
+	PregelLoading PregelJobState = "loading"
```
Comment (nikita-vanyasin) [inline database_pregel.go:75]:
```suggestion
	PregelJobStateLoading PregelJobState = "loading"
```

File: database_pregel.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+	// PregelJobStateInError - The execution is in an error state. This can be caused by DB-Servers being not reachable or being non-responsive.
+	// The execution might recover later, or switch to "canceled" if it was not able to recover successfully.
+	PregelJobStateInError PregelJobState = "in error"
+	// PregelJobStateRecovering - (currently unused): The execution is actively recovering and switches back to running if the recovery is successful.
+	PregelJobStateRecovering PregelJobState = "recovering"
+)
+
+type PregelJob struct {
+	// The ID of the Pregel job, as a string.
+	Id string `json:"id"`
```
Comment (nikita-vanyasin) [inline database_pregel.go:96]:
```suggestion
	ID string `json:"id"`
```

File: database_pregel.go [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+	// The total runtime of the execution up to now (if the execution is still ongoing).
+	TotalRuntime float64 `json:"totalRuntime,omitempty"`
+	// The startup runtime of the execution. The startup time includes the data loading time and can be substantial.
+	StartupTime float64 `json:"startupTime,omitempty"`
+	// The algorithm execution time. Is shown when the computation started.
+	ComputationTime float64 `json:"computationTime,omitempty"`
+	// The time for storing the results if the job includes results storage. Is shown when the storing started.
+	StorageTime float64 `json:"storageTime,omitempty"`
+	// Computation time of each global super step. Is shown when the computation started.
+	GssTimes []float64 `json:"gssTimes,omitempty"`
```
Comment (nikita-vanyasin) [inline database_pregel.go:120]:
```suggestion
	GSSTimes []float64 `json:"gssTimes,omitempty"`
```
go guidelines naming ^

File: database_pregel.go [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+// AggregatedStatus The aggregated details of the full Pregel run. The values are totals of all the DB-Server.
+type AggregatedStatus struct {
+	// The time at which the status was measured.
+	TimeStamp time.Time `json:"timeStamp,omitempty"`
+	// The status of the in memory graph.
+	GraphStoreStatus *GraphStoreStatus `json:"graphStoreStatus,omitempty"`
+	//  Information about the global supersteps.
+	AllGssStatus *AllGssStatus `json:"allGssStatus,omitempty"`
```
Comment (nikita-vanyasin) [inline database_pregel.go:151]:
```suggestion
	AllGSSStatus *AllGSSStatus `json:"allGssStatus,omitempty"`
```

File: database_pregel.go [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+// AllGssStatus Information about the global supersteps.
+type AllGssStatus struct {
+	//  A list of objects with details for each global superstep.
+	Items []GssStatus `json:"items,omitempty"`
+}
+
+// GssStatus Information about the global superstep
+type GssStatus struct {
```
Comment (nikita-vanyasin) [inline database_pregel.go:173]:
```suggestion
type GSSStatus struct {
```

File: test/pregel_test.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+	"github.com/arangodb/go-driver"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestCreatePregelJob(t *testing.T) {
+	ctx := context.Background()
+	c := createClientFromEnv(t, true)
+	skipBelowVersion(c, "3.10", t)
```
Comment (nikita-vanyasin) [inline test/pregel_test.go:35]: Probably these tests should executed only at EE versions? The Arangodb docs says pregel works only with smart graphs, which are available only in EE

File: test/pregel_test.go [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+		Algorithm: driver.PregelAlgorithmPageRank,
+		GraphName: g.Name(),
+	})
+	require.Nilf(t, err, "Failed to start Pregel job: %s", describe(err))
+	require.NotEmpty(t, jobId, "JobId is empty")
+
+	job, err := db.GetJob(ctx, jobId)
+	require.Nilf(t, err, "Failed to get job: %s", describe(err))
+	require.Equal(t, jobId, job.Id, "JobId mismatch")
+	//require.Equal(t, driver.PregelAlgorithmPageRank, job.Algorithm, "Algorithm mismatch")
```
Comment (nikita-vanyasin) [inline test/pregel_test.go:67]:
```suggestion
```

**Review by nikita-vanyasin — APPROVED** [review]
LGTM with a few comments

File: test/pregel_test.go (hunk 8 above)
Comment (jwierzbo) [reply test/pregel_test.go:35]: it works on Community version as well :)

File: test/pregel_test.go (hunk 9 above)
Comment (jwierzbo) [reply test/pregel_test.go:67]: waiting for Jira ticket - there is a bug in the Algorithm field

File: database_pregel.go (hunk 1 above)
Comment (jwierzbo) [reply database_pregel.go:65]: this field is general as general are alg-spec options..

**Pull request merged**